[Merged by Bors] - refactor(algebra/big_operators, *): specialize `finset.prod_bij` to `fintype.prod_equiv`

Vierkantor opened the pull request:
Often we want to apply `finset.prod_bij` to the case where the product is taken over `finset.univ` and the bijection is already a bundled `equiv`. This PR adds `fintype.prod_equiv` as a specialization for exactly these cases, filling in most of the arguments already.

---

- [x] depends on: #7083 

[![Open in Gitpod](https://gitpod.io/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
refactor(algebra/big_operators, *): specialize `finset.prod_bij` to `fintype.prod_equiv`

Often we want to apply `finset.prod_bij` to the case where the product is taken
over `finset.univ` and the bijection is already a bundled `equiv`. This PR adds
`fintype.prod_equiv` as a specialization for exactly these cases, filling in
most of the arguments already.

I still want to go through the usages of `prod_bij` to see where the new
`prod_equiv` can be usefully applied.

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1276,6 +1276,26 @@ end comm_group_with_zero
+automatically fills in most arguments."]
+lemma prod_equiv {α β M : Type*} [fintype α] [fintype β] [comm_monoid M]
+  (e : α ≃ β) (f : α → M) (g : β → M) (h : ∀ x, f x = g (e x)) :
+  ∏ x : α, f x = ∏ x : β, g x :=
```
I'm curious whether this is more or less useful as 
```suggestion
  (e : α ≃ β) (f : α → M) (g : β → M) :
  ∏ x : α, g (e x) = ∏ x : β, g x :=
```

Vierkantor replied:
I'm still going through all the usages (mostly waiting for code to compile...), but it looks like there is only one instance where `h = refl` would work out of about 10. So I don't think the change is worth it.

eric-wieser replied:
I guess this comes down to whether the lemma is intended to be used via apply or via `rw`.

eric-wieser replied:
My request is stupid, the lemma I'm describing already exists and is called `equiv.prod_comp`!

Vierkantor committed:
Add a few `simp` lemmas for equivs

Vierkantor committed:
Apply the obvious `prod_bij` -> `prod_equiv` changes

I think I'll leave it at these for now, since most other instances are sums
over finsets, or use `prod_bij_ne_one` which doesn't seem to be able to be
decomplicated much.

Vierkantor commented on src/data/equiv/basic.lean:
```diff
@@ -129,6 +129,9 @@ def simps.inv_fun (e : α ≃ β) : β → α := e.symm
 
 initialize_simps_projections equiv (to_fun → apply, inv_fun → symm_apply)
 
+-- Generate the `simps` projections for previously defined equivs.
```
I can also split the changes in this file into a new PR if you prefer. (Overall there is a big defecit of `simps`es in `data/equiv`.)

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1276,6 +1276,26 @@ end comm_group_with_zero
+  (λ x _, mem_univ (e x))
+  (λ x _, h x)
+  (λ x x' _ _ h, e.injective h)
+  (λ y _, ⟨e.symm y, mem_univ _, (e.apply_symm_apply y).symm⟩)
```
A fair number of the cases below seem to use `equiv.of_bijective`, which uses the axiom of choice. Why not have a version of this that takes `function.bijective` to save the trouble of using `equiv.of_bijective`, and as a bonus avoid the need for a computable inverse?
```suggestion
/-- `fintype.prod_bijective` is a variant of `finset.prod_bij` that accepts `function.bijective`. -/
@[to_additive "`fintype.sum_equiv` is a variant of `finset.sum_bij` that accepts
`function.bijective`"]
lemma prod_bijective {α β M : Type*} [fintype α] [fintype β] [comm_monoid M]
  (e : α → β) (he : function.bijective e) (f : α → M) (g : β → M) (h : ∀ x, f x = g (e x)) :
  ∏ x : α, f x = ∏ x : β, g x :=
prod_bij
  (λ x _, e x)
  (λ x _, mem_univ (e x))
  (λ x _, h x)
  (λ x x' _ _ h, he.injective h)
  (λ y _, (he.surjective y).imp $ λ a h, ⟨mem_univ _, h.symm⟩)

/-- `fintype.prod_equiv` is a specialization of `finset.prod_bij` that
automatically fills in most arguments. -/
@[to_additive "`fintype.sum_equiv` is a specialization of `finset.sum_bij` that
automatically fills in most arguments."]
lemma prod_equiv {α β M : Type*} [fintype α] [fintype β] [comm_monoid M]
  (e : α ≃ β) (f : α → M) (g : β → M) (h : ∀ x, f x = g (e x)) :
  ∏ x : α, f x = ∏ x : β, g x :=
prod_bijective e e.bijective f g h
```

eric-wieser commented on src/algebra/polynomial/group_ring_action.lean:
```diff
@@ -98,10 +98,9 @@ theorem prod_X_sub_smul.eval (x : R) : (prod_X_sub_smul G R x).eval x = 0 :=
-  (λ g' _, by rw [of_quotient_stabilizer_smul, smul_sub, polynomial.smul_X, polynomial.smul_C])
-  (λ _ _ _ _ H, (mul_action.bijective g).1 H)
-  (λ g' _, ⟨g⁻¹ • g', finset.mem_univ _, by rw [smul_smul, mul_inv_self, one_smul]⟩)
+(smul_prod _ _ _ _ _).trans $ fintype.prod_equiv (equiv.of_bijective _ (mul_action.bijective g)) _ _
```
```suggestion
(smul_prod _ _ _ _ _).trans $ fintype.prod_bijective (mul_action.bijective g) _ _
```

eric-wieser commented on src/data/fintype/card.lean:
```diff
@@ -250,8 +249,7 @@ by simpa using fintype.sum_pow_mul_eq_add_pow (fin n) a b
   ∏ i, g (f i) = ∏ i, g i :=
-prod_bij (λ i hi, f i) (λ i hi, mem_univ _) (λ i hi, rfl) (λ i j _ _ h, hf.1 h) $
-  λ i hi, (hf.2 i).imp $ λ j hj, ⟨mem_univ _, hj.symm⟩
+fintype.prod_equiv (equiv.of_bijective f hf) _ _ (λ x, rfl)
```
```suggestion
fintype.prod_bijective h hf _ _ (λ x, rfl)
```

eric-wieser commented on src/linear_algebra/determinant.lean:
```diff
@@ -170,18 +166,13 @@ instance : is_monoid_hom (det : matrix n n R → R) :=
-    { intros i _, use σ⁻¹ i, finish } },
-  { intros σ σ' _ _ h, simp at h, assumption },
-  { intros σ _, use σ⁻¹, finish }
+  refine fintype.sum_equiv (inv_involutive.to_equiv _) _ _ _,
```
```suggestion
  refine fintype.sum_bijective (inv_involutive.bijective) _ _ _,
```

eric-wieser reviewed (commented):
Below changes are untested, but might work after introducing the new definition

Vierkantor committed:
Add `prod_bijective`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

Vierkantor committed:
Use `fintype.prod_bijective` in a couple of places

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1276,6 +1276,34 @@ end comm_group_with_zero
+
+/-- `fintype.prod_bijective` is a variant of `finset.prod_bij` that accepts `function.bijective`. -/
+@[to_additive "`fintype.sum_equiv` is a variant of `finset.sum_bij` that accepts
+`function.bijective`"]
```
```suggestion
/-- `fintype.prod_bijective` is a variant of `finset.prod_bij` that accepts `function.bijective`.

See `function.bijective.prod_comp` for a version without `h`. -/
@[to_additive "`fintype.sum_equiv` is a variant of `finset.sum_bij` that accepts
`function.bijective`.

See `function.bijective.sum_comp` for a version without `h`. "]
```

urkud replied:
We already have `function.bijective.prod_comp`, do we need one more version?

eric-wieser replied:
It sounds like the presence of the `h` argument makes the lemma a lot more usable, as demonstrated by the rest of the diff

eric-wieser commented on src/algebra/big_operators/basic.lean:
```diff
@@ -1276,6 +1276,34 @@ end comm_group_with_zero
+/-- `fintype.prod_equiv` is a specialization of `finset.prod_bij` that
+automatically fills in most arguments. -/
+@[to_additive "`fintype.sum_equiv` is a specialization of `finset.sum_bij` that
+automatically fills in most arguments."]
```
```suggestion
/-- `fintype.prod_equiv` is a specialization of `finset.prod_bij` that
automatically fills in most arguments.

See `equiv.prod_comp` for a version without `h`. 
-/
@[to_additive "`fintype.sum_equiv` is a specialization of `finset.sum_bij` that
automatically fills in most arguments.

See `equiv.sum_comp` for a version without `h`. 
"]
```

eric-wieser reviewed (approved):
Looks good now, thanks!

I had no idea that `equiv.prod_comp` existed until I reviewed this PR - so I think we should mention it in the docstrings, as suggested below.

bors d+

Vierkantor committed:
Refer to `function.bijective.{sum,prod}_comp` in docstrings

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>
